crates/btcio/src/reader/query.rs

```diff
@@ -279,7 +299,116 @@ async fn poll_for_new_blocks<R: Reader>(
+    pub fn new(mut chainstate_update_timeout: Duration, max_wait_time: Duration) -> Self {
+        if max_wait_time < chainstate_update_timeout {
+            warn!(
+                ?max_wait_time,
+                ?chainstate_update_timeout,
+                "checkpointconfig: max_wait_time smaller than chainstate_update_timeout, clamping chainstate_update_tomeout to be the same"
+            );
+            chainstate_update_timeout = max_wait_time;
+        }
+        Self {
+            chainstate_update_timeout,
+            max_wait_time,
+        }
+    }
+
+    pub fn chainstate_update_timeout(&self) -> &Duration {
+        &self.chainstate_update_timeout
+    }
+
+    pub fn max_wait_time(&self) -> &Duration {
+        &self.max_wait_time
+    }
+}
+
+impl Default for CheckpointWaitConfig {
+    fn default() -> Self {
+        Self {
+            chainstate_update_timeout: Duration::from_secs(30),
+            max_wait_time: Duration::from_secs(300), // 5 minutes
+        }
+    }
+}
+
+/// Watches the status channel for chainstate which has the finalized state matching as referred by
+/// checkpoint. If it is found, returns `true`, otherwise returns `false`. `false` is returned when
+/// there is a different chainstate than indicated by checkpoint for given slot.
+async fn wait_until_checkpoint_accepted_or_rejected<R: Reader>(
+    ctx: &ReaderContext<R>,
+    checkpoint: &SignedCheckpoint,
+) -> anyhow::Result<bool> {
+    let l2_commt = checkpoint.checkpoint().batch_info().l2_range.1;
+    let config = &ctx.checkpoint_config;
+
+    info!("Waiting for checkpoint to be accepted");
+    let mut rx = ctx.status_channel.subscribe_chain_sync();
+
+    let start_time = time::Instant::now();
+
+    loop {
+        if start_time.elapsed() > config.max_wait_time {
+            error!("Maximum wait time exceeded waiting for checkpoint acceptance");
+            return Err(anyhow!(
+                "reader:query:wait_for_checkpoint_accept: Maximum wait time of {:?}s exceeded",
+                config.max_wait_time
+            ));
+        } else if let Some(chstate_update) = (*rx.borrow_and_update()).clone() {
+            let tip_commitment = chstate_update.new_status().tip;
+            let same_slot = l2_commt.slot() == tip_commitment.slot();
+            let same_blkid = l2_commt.blkid() == tip_commitment.blkid();
+
+            if same_slot && same_blkid {
+                return Ok(true);
+            } else if same_slot && !same_blkid {
+                warn!("Checkpoint not accepted: different block id for same slot finalized");
+                return Ok(false);
+            } else if l2_commt.slot() < tip_commitment.slot() {
+                // We've been looking for stale checkpoint to be accepted.
+                warn!("Checkpoint not accepted: higher slot finalized than the slot waited for");
+                return Ok(false);
+            }
+        }
+
+        wait_for_chainstate_change(config, &mut rx).await?;
```

Comment:
This control flow is really convoluted and indirect still.

The call to `wait_for_chainstate_change` has the information figure out what condition it is, but it discards it and we only use it as a way for us to signal to do another iteration on the loop.  Then in the next iteration we go and perform the checks to decide if we should either propagate up the timeout, get the value back out of the watch channel a second time and do checks, or just wait again.

It should return some thing like

```rs
enum QueryEvent {
  Update(ChainSyncStatusUpdate),
  Timeout,
  Exit // or make it return `Some(QueryEvent)`
}
```

and then pattern match on that to figure out what to do, then just wait on the function again.